text/3374-cargo-target-features.md

```diff
@@ -0,0 +1,177 @@
+- Feature Name: `cargo_target_features`
+- Start Date: 2023-01-20
+- RFC PR: [rust-lang/rfcs#3374](https://github.com/rust-lang/rfcs/pull/3374)
+- Tracking Issue: [rust-lang/cargo#0000](https://github.com/rust-lang/cargo/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This adds a new `enable-features` field to [Cargo Targets](https://doc.rust-lang.org/cargo/reference/cargo-targets.html#configuring-a-target) which forces a [feature](https://doc.rust-lang.org/cargo/reference/features.html) to be enabled if the target is being built.
+
+# Motivation
+[motivation]: #motivation
+
+There are several situations where a non-library target must have a particular feature enabled for it to work correctly.
+Although it is possible to manually enable these features on the command-line, this can make it awkward to use.
+This RFC adds a mechanism to make it easier to work with these targets.
+
+Some example use cases are:
+
+1. Binaries that need additional dependencies like command-line processing libraries, or logging functionality.
+2. Examples illustrating how to use a library with a specific feature enabled.
+3. Benchmarks which require a separate benchmarking library, but you don't want to pay the cost of building that library when not running benchmarks.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The `enable-features` field can be added to a target in `Cargo.toml` to specify features or dependencies that will be automatically enabled whenever the target is being built.
+For example:
+
+```toml
+[package]
+name = "myproject"
+version = "0.1.0"
+edition = "2021"
+
+[[example]]
+name = "fetch-http"
+enable-features = ["networking"]
+
+[[bin]]
+name = "myproject"
+enable-features = ["dep:clap"]
```

Comment:
In #3020, I have the following line:

> For a target, specifying `lib-features` does not implicity activate the default features of the library dependency. If needed, the target can specify `default` in the list of values for `lib-features`.

And there were some concerns about that as described in https://github.com/rust-lang/rfcs/pull/3020#issuecomment-862381779

The original motivation for having that line is to allow the library and binary to have separate dependencies. There are use cases where the user use the same `cool` name for a library and wants to have a cli tool with the same `cool` name but with different dependencies. (ex: cli tool only does project generation to use the library). In this scenario, when the end user runs `cargo install cool`, the unneeded dependencies present in `default` are also built.

To solve this issue, I would propose what I had been thinking last on #3020.

```toml
[[bin]]
name = "foo"
enable = { features = ["dep:clap"], no-default-features = true }

[[bin]]
name = "bar"
enable = ["dep:clap"] # shorthand

[dependencies]
clap = { version = "4.0.26", optional = true }
serde = { version = "1.0.152", optional = true }

[features]
default = ["dep:serde"]
```

If I were to run `cargo install cool --bin foo`, default features will not be activated:

```
 Compiling clap v4.0.26
Installing /bin/foo
```

If I were to run `cargo install cool` (installing both binaries), the default features will be activated because of `bar` target:

```
 Compiling clap v4.0.26
 Compiling serde v1.0.152
Installing /bin/foo
Installing /bin/bar
```

Reply:
Thanks for the suggestion! I have added it as a possible extension to consider.